File: app/src/main/java/com/asfoundation/wallet/repository/WalletRepository.java
package com.asfoundation.wallet.repository;

import com.asfoundation.wallet.analytics.AmplitudeAnalytics;
import com.asfoundation.wallet.analytics.AnalyticsSetup;
import com.asfoundation.wallet.entity.Wallet;
import com.asfoundation.wallet.service.AccountKeystoreService;
import com.asfoundation.wallet.service.WalletBalanceService;
import io.reactivex.Completable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import java.math.BigDecimal;
import org.jetbrains.annotations.NotNull;

public class WalletRepository implements WalletRepositoryType {

  private final PreferencesRepositoryType preferencesRepositoryType;
  private final AccountKeystoreService accountKeystoreService;
  private final WalletBalanceService walletBalanceService;
  private final Scheduler networkScheduler;
  private final AnalyticsSetup analyticsSetUp;
  private final AmplitudeAnalytics amplitudeAnalytics;

  public WalletRepository(PreferencesRepositoryType preferencesRepositoryType,
      AccountKeystoreService accountKeystoreService, WalletBalanceService walletBalanceService,
      Scheduler networkScheduler, AnalyticsSetup analyticsSetUp,
      AmplitudeAnalytics amplitudeAnalytics) {
    this.preferencesRepositoryType = preferencesRepositoryType;
    this.accountKeystoreService = accountKeystoreService;
    this.walletBalanceService = walletBalanceService;
    this.networkScheduler = networkScheduler;
    this.analyticsSetUp = analyticsSetUp;
    this.amplitudeAnalytics = amplitudeAnalytics;
  }

  @Override public Single<Wallet[]> fetchWallets() {
    return accountKeystoreService.fetchAccounts();
  }

  @Override public Single<Wallet> findWallet(String address) {
    return fetchWallets().flatMap(accounts -> {
      for (Wallet wallet : accounts) {
        if (wallet.sameAddress(address)) {
          return Single.just(wallet);
        }
      }
      return null;
    });
  }

  @Override public Single<Wallet> createWallet(String password) {
    return accountKeystoreService.createAccount(password);
  }

  @Override
  public Single<Wallet> restoreKeystoreToWallet(String store, String password, String newPassword) {
    return accountKeystoreService.restoreKeystore(store, password, newPassword);
  }

  @Override public Single<Wallet> restorePrivateKeyToWallet(String privateKey, String newPassword) {
    return accountKeystoreService.restorePrivateKey(privateKey, newPassword);
  }

  @Override
  public Single<String> exportWallet(String address, String password, String newPassword) {
    return accountKeystoreService.exportAccount(address, password, newPassword);
  }

  @Override public Completable deleteWallet(String address, String password) {
    return accountKeystoreService.deleteAccount(address, password);
  }

  @Override public Completable setDefaultWallet(String address) {
    return Completable.fromAction(() -> {
      analyticsSetUp.setUserId(address);
      amplitudeAnalytics.setUserId(address);
      preferencesRepositoryType.setCurrentWalletAddress(address);
    });
  }

  @Override public Single<Wallet> getDefaultWallet() {
    return Single.fromCallable(() -> {
      String currentWalletAddress = preferencesRepositoryType.getCurrentWalletAddress();
      if (currentWalletAddress == null) {
        throw new WalletNotFoundException();
      } else {
        return currentWalletAddress;
      }
    })
        .flatMap(this::findWallet);
  }

  @NotNull @Override public Single<BigDecimal> getEthBalanceInWei(String address) {
    return walletBalanceService.getWalletBalance(address)
        .map(walletBalance -> new BigDecimal(walletBalance.getEth()))
        .subscribeOn(networkScheduler);
  }

  @NotNull @Override public Single<BigDecimal> getAppcBalanceInWei(String address) {
    return walletBalanceService.getWalletBalance(address)
        .map(walletBalance -> new BigDecimal(walletBalance.getAppc()))
        .subscribeOn(networkScheduler);
  }
}

File: app/src/main/java/com/asfoundation/wallet/repository/WalletRepository.kt
package com.asfoundation.wallet.repository

import android.content.SharedPreferences
import com.asfoundation.wallet.analytics.AnalyticsSetup
import com.asfoundation.wallet.entity.Wallet
import com.asfoundation.wallet.interact.rx.operator.Operators
import com.asfoundation.wallet.recover.result.RestoreResult
import com.asfoundation.wallet.recover.result.SuccessfulRestore
import com.asfoundation.wallet.service.AccountKeystoreService
import io.reactivex.*
import it.czerwinski.android.hilt.annotations.BoundTo
import repository.CommonsPreferencesDataSource
import repository.CommonsPreferencesDataSource.Companion.CURRENT_ACCOUNT_ADDRESS_KEY
import javax.inject.Inject

@BoundTo(supertype = WalletRepositoryType::class)
class WalletRepository @Inject constructor(
  private val commonsPreferencesDataSource: CommonsPreferencesDataSource,
  private val accountKeystoreService: AccountKeystoreService,
  private val analyticsSetUp: AnalyticsSetup,
  private val passwordStore: PasswordStore
) : WalletRepositoryType {

  override fun fetchWallets(): Single<Array<Wallet>> {
    return accountKeystoreService.fetchAccounts()
  }

  override fun findWallet(address: String): Single<Wallet> {
    return fetchWallets().map { accounts ->
      for (wallet in accounts) {
        if (wallet.hasSameAddress(address)) {
          return@map wallet
        }
      }
      throw NullPointerException("No wallets found")
    }
  }

  override fun createWallet(password: String): Single<Wallet> {
    return accountKeystoreService.createAccount(password)
  }

  override fun restoreKeystoreToWallet(store: String, password: String,
                                       newPassword: String): Single<RestoreResult> {
    return accountKeystoreService.restoreKeystore(store, password, newPassword)
        .flatMap { restoreResult ->
          var savePasswordCompletable = Completable.complete()
          if (restoreResult is SuccessfulRestore) {
            savePasswordCompletable = savePassword(restoreResult.address, newPassword)
          }
          return@flatMap savePasswordCompletable
              .andThen(Single.just(restoreResult))
        }
  }

  override fun restorePrivateKeyToWallet(privateKey: String?,
                                         newPassword: String): Single<RestoreResult> {
    return accountKeystoreService.restorePrivateKey(privateKey, newPassword)
        .flatMap { restoreResult ->
          var savePasswordCompletable = Completable.complete()
          if (restoreResult is SuccessfulRestore) {
            savePasswordCompletable = savePassword(restoreResult.address, newPassword)
          }
          return@flatMap savePasswordCompletable
              .andThen(Single.just(restoreResult))
        }
  }

  override fun exportWallet(address: String, password: String,
                            newPassword: String?): Single<String> {
    return accountKeystoreService.exportAccount(address, password, newPassword)
  }

  override fun deleteWallet(address: String, password: String): Completable {
    return accountKeystoreService.deleteAccount(address, password)
  }

  override fun setDefaultWallet(address: String): Completable {
    return Completable.fromAction {
      analyticsSetUp.setUserId(address)
      commonsPreferencesDataSource.setCurrentWalletAddress(address)
    }
  }

  override fun getDefaultWallet(): Single<Wallet> {
    return Single.fromCallable { getDefaultWalletAddress() }
        .flatMap { address -> findWallet(address) }
  }

  private fun getDefaultWalletAddress(): String {
    val currentWalletAddress = commonsPreferencesDataSource.getCurrentWalletAddress()
    return currentWalletAddress ?: throw WalletNotFoundException()
  }

  private fun emitWalletAddress(emitter: ObservableEmitter<String>) {
    try {
      val walletAddress = getDefaultWalletAddress()
      emitter.onNext(walletAddress)
    } catch (e: WalletNotFoundException) {
      emitter.tryOnError(e)
    }
  }

  override fun observeDefaultWallet(): Observable<Wallet> {
    return Observable.create(
        ObservableOnSubscribe { emitter: ObservableEmitter<String> ->
          val listener =
              SharedPreferences.OnSharedPreferenceChangeListener { _, key: String ->
                if (key == CURRENT_ACCOUNT_ADDRESS_KEY) {
                  emitWalletAddress(emitter)
                }
              }
          emitter.setCancellable { commonsPreferencesDataSource.removeChangeListener(listener) }
          emitWalletAddress(emitter)
          commonsPreferencesDataSource.addChangeListener(listener)
        } as ObservableOnSubscribe<String>)
        .flatMapSingle { address -> findWallet(address) }
  }

  override fun savePassword(address: String, password: String): Completable {
    return passwordStore.setPassword(address, password)
          .onErrorResumeNext { throwable ->
            deleteWallet(address, password)
                .lift(Operators.completableErrorProxy(throwable))
          }
  }
}